File: src/types/provider.ts
/**
 * 供应商管理系统类型定义
 */

/**
 * 供应商信息
 */
export interface Provider {
  /** 供应商唯一标识（如 "duckcoding"） */
  id: string;
  /** 供应商名称（用于显示） */
  name: string;
  /** 供应商官网地址 */
  website_url: string;
  /** API 地址（可选，优先于 website_url 用于 API 调用） */
  api_address?: string;
  /** 用户ID */
  user_id: string;
  /** 访问令牌 */
  access_token: string;
  /** 用户名（可选） */
  username?: string;
  /** 是否为默认供应商 */
  is_default: boolean;
  /** 创建时间（Unix timestamp） */
  created_at: number;
  /** 更新时间（Unix timestamp） */
  updated_at: number;
  /** 签到配置（可选） */
  checkin_config?: CheckinConfig;
}

/**
 * 签到配置
 */
export interface CheckinConfig {
  /** 是否启用自动签到 */
  enabled: boolean;
  /** 签到 API 端点（相对路径，如 "/api/user/checkin"） */
  endpoint: string;
  /** 签到时间范围 - 开始小时（0-23，默认 0） */
  checkin_hour_start?: number;
  /** 签到时间范围 - 结束小时（0-23，默认 0）；start==end 或 start>end 时为全天 */
  checkin_hour_end?: number;
  /** 下次计划签到时间（Unix timestamp），由后端调度器生成 */
  next_checkin_at?: number;
  /** 最后签到时间（Unix timestamp） */
  last_checkin_at?: number;
  /** 最后签到状态 */
  last_checkin_status?: 'success' | 'failed';
  /** 最后签到消息 */
  last_checkin_message?: string;
  /** 累计签到次数 */
  total_checkins?: number;
  /** 累计获得额度 */
  total_quota?: number;
}

/**
 * 签到响应
 */
export interface CheckinResponse {
  success: boolean;
  message?: string;
  data?: {
    quota_awarded?: number;
    checkin_date?: string;
    stats?: {
      checked_in_today: boolean;
      checkin_count: number;
      total_checkins: number;
      total_quota: number;
      records?: Array<{
        checkin_date: string;
        quota_awarded: number;
      }>;
    };
  };
}

/**
 * API 地址信息
 */
export interface ApiInfo {
  /** API 地址 URL */
  url: string;
  /** 地址描述 */
  description: string;
}

/**
 * 供应商存储结构
 */
export interface ProviderStore {
  /** 数据版本 */
  version: number;
  /** 供应商列表 */
  providers: Provider[];
  /** 当前激活的供应商ID */
  active_provider_id?: string;
  /** 最后更新时间（Unix timestamp） */
  updated_at: number;
}

/**
 * 供应商配置表单数据（暂未使用，保留给 UI 组件）
 */
export interface _ProviderFormData {
  /** 供应商名称 */
  name: string;
  /** 官网地址 */
  website_url: string;
  /** 用户ID */
  user_id: string;
  /** 访问令牌 */
  access_token: string;
  /** 用户名（可选） */
  username?: string;
}

/**
 * 供应商验证结果
 */
export interface ProviderValidationResult {
  /** 是否验证成功 */
  success: boolean;
  /** 从 API 获取的用户名（用于确认身份） */
  username?: string;
  /** 错误消息（验证失败时） */
  error?: string;
}

File: src-tauri/src/models/provider.rs
// Provider Configuration Models
//
// 供应商配置数据模型

use serde::{Deserialize, Serialize};

/// 签到配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckinConfig {
    /// 是否启用自动签到
    pub enabled: bool,
    /// 签到 API 端点
    pub endpoint: String,
    /// 签到时间范围 - 开始小时 (0-23，默认 0)
    #[serde(default)]
    pub checkin_hour_start: u8,
    /// 签到时间范围 - 结束小时 (0-23，默认 0)
    /// start == end 或 start > end 时视为全天 (0-23)
    #[serde(default)]
    pub checkin_hour_end: u8,
    /// 下次计划签到时间 (Unix timestamp)，由调度器生成随机时间
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_checkin_at: Option<i64>,
    /// 最后签到时间
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_checkin_at: Option<i64>,
    /// 最后签到状态
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_checkin_status: Option<String>,
    /// 最后签到消息
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_checkin_message: Option<String>,
    /// 累计签到次数
    #[serde(default)]
    pub total_checkins: u32,
    /// 累计获得额度
    #[serde(default)]
    pub total_quota: i64,
}

impl CheckinConfig {
    /// 获取有效的签到时间范围 (start_hour, end_hour)
    /// start < end 时返回实际范围，否则返回全天 (0, 23)
    pub fn effective_range(&self) -> (u8, u8) {
        if self.checkin_hour_start < self.checkin_hour_end {
            (self.checkin_hour_start, self.checkin_hour_end)
        } else {
            (0, 23)
        }
    }
}

impl Default for CheckinConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            endpoint: "/api/user/checkin".to_string(),
            checkin_hour_start: 0,
            checkin_hour_end: 0,
            next_checkin_at: None,
            last_checkin_at: None,
            last_checkin_status: None,
            last_checkin_message: None,
            total_checkins: 0,
            total_quota: 0,
        }
    }
}

/// 供应商配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Provider {
    /// 唯一标识
    pub id: String,
    /// 供应商名称（如 DuckCoding）
    pub name: String,
    /// 官网地址
    pub website_url: String,
    /// API 地址（可选，优先于 website_url 用于 API 调用）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_address: Option<String>,
    /// 用户ID
    pub user_id: String,
    /// 系统访问令牌
    pub access_token: String,
    /// 用户名（可选，用于确认）
    pub username: Option<String>,
    /// 是否为默认供应商
    pub is_default: bool,
    /// 创建时间
    pub created_at: i64,
    /// 更新时间
    pub updated_at: i64,
    /// 签到配置
    #[serde(skip_serializing_if = "Option::is_none")]
    pub checkin_config: Option<CheckinConfig>,
}

/// 供应商存储结构
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderStore {
    /// 数据版本
    pub version: u32,
    /// 供应商列表
    pub providers: Vec<Provider>,
    /// 最后更新时间
    pub updated_at: i64,
}

impl Default for ProviderStore {
    fn default() -> Self {
        let now = chrono::Utc::now().timestamp();
        Self {
            version: 1,
            providers: vec![Provider {
                id: "duckcoding".to_string(),
                name: "DuckCoding".to_string(),
                website_url: "https://duckcoding.com".to_string(),
                api_address: Some("https://jp.duckcoding.com".to_string()),
                user_id: String::new(),
                access_token: String::new(),
                username: None,
                is_default: true,
                created_at: now,
                updated_at: now,
                checkin_config: None,
            }],
            updated_at: now,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_provider_store() {
        let store = ProviderStore::default();
        assert_eq!(store.version, 1);
        assert_eq!(store.providers.len(), 1);
        assert_eq!(store.providers[0].id, "duckcoding");
        assert_eq!(store.providers[0].name, "DuckCoding");
        assert!(store.providers[0].is_default);
    }

    #[test]
    fn test_provider_serialization() {
        let provider = Provider {
            id: "test".to_string(),
            name: "Test Provider".to_string(),
            website_url: "https://test.com".to_string(),
            api_address: Some("https://api.test.com".to_string()),
            user_id: "12345".to_string(),
            access_token: "token123".to_string(),
            username: Some("testuser".to_string()),
            is_default: false,
            created_at: 1234567890,
            updated_at: 1234567890,
            checkin_config: None,
        };

        let json = serde_json::to_string(&provider).unwrap();
        let deserialized: Provider = serde_json::from_str(&json).unwrap();

        assert_eq!(deserialized.id, provider.id);
        assert_eq!(deserialized.name, provider.name);
        assert_eq!(deserialized.api_address, provider.api_address);
        assert_eq!(deserialized.username, provider.username);
    }
}
